src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -1161,6 +1170,26 @@ def __create_internal(self, job, data):
                     {'prefix': self._config.datastore_prefix}
                 )
 
+    @private
+    def revoke_acme_certificate(self, id):
+
+        certificate = self.middleware.call_sync('certificate._get_instance', id)
+
+        if certificate.get('acme'):
+            client, key = self.get_acme_client_and_key(certificate['acme']['directory'], True)
+
+            try:
+                client.revoke(
+                    jose.ComparableX509(
+                        crypto.load_certificate(crypto.FILETYPE_PEM, certificate['certificate'])
+                    ),
+                    0
+                )
+            except (errors.ClientError, messages.Error) as e:
+                return str(e)
```

Comment:
Why not `raise CallError`?

Reply:
We do raise it in delete if force is not True below

Reply:
I see, but why do we return a string instead of raising an exception? This can confuse people. I understand why C people do it when they work on middleware, it's their habit, but why should we?